chord/chordServerImpl.go

```diff
@@ -54,14 +56,111 @@ func (chordNode *ChordNode) FindSuccessor(ctx context.Context, messageIDPtr *ID)
 // n.notify(n0)
 //	 if (predecessor is nil or n0 is_in (predecessor; n))
 //	 	predecessor = n0;
-func (chordNode *ChordNode) Notify(ctx context.Context, n0 *Node) (*empty.Empty, error) {
+func (chordNode *ChordNode) Notify(ctx context.Context, n0Ptr *Node) (*empty.Empty, error) {
 	emptyPtr := &empty.Empty{}
-	if n0 == nil {
+
+	switch {
+	case n0Ptr == nil:
 		return emptyPtr, errors.New("trying to notify nil node")
+	case n0Ptr.Id == chordNode.node.Id:
+		return emptyPtr, errors.New(fmt.Sprintf("node %s is notifying itself", chordNode.node.Id))
 	}
-	pred, doesExist := chordNode.getPredecessor()
-	if (!doesExist && n0.Id != chordNode.node.Id) || isBetweenTwoNodesExclusive(pred, *n0, chordNode.node) {
-		chordNode.setPredecessor(n0)
+
+	pred, doesPredExist := chordNode.getPredecessor()
+	if !doesPredExist || isBetweenTwoNodesExclusive(pred.Id, n0Ptr.Id, chordNode.node.Id) {
+		chordNode.setPredecessor(n0Ptr)
+		if doesPredExist {
+			_, err := chordNode.TransferKeys(context.Background(), &TransferKeysRequest{
+				FromId: pred.Id,
+				ToNode: n0Ptr,
+			})
+			if err != nil {
+				log.WithFields(logrus.Fields{
+					"fromID": pred.Id,
+					"toID":   n0Ptr.Id,
+					"err":    err,
+				}).Warn("TransferKeys failed")
+			}
+		}
+	}
+
+	return emptyPtr, nil
+}
+
+func (chordNode *ChordNode) Put(ctx context.Context, keyValPtr *FileRecordWithKeyword) (*empty.Empty, error) {
+	emptyPtr := &empty.Empty{}
+
+	if keyValPtr == nil {
+		return emptyPtr, nilError("FileRecordWithKeyword")
+	}
+
+	err := chordNode.hashTable.PutOrAppendOne(keyValPtr.Keyword, keyValPtr.Val)
+
+	return emptyPtr, err
+}
+
+// Get implements RPC method Get (Key) returns (Val);
+// It returns a list of documents that were stored under this keyword
+func (chordNode *ChordNode) Get(ctx context.Context, messageKeyPtr *Key) (*Val, error) {
+	if messageKeyPtr == nil {
+		return nil, nilError("Key")
+	}
+
+	chordNode.hashTable.RLock()
+	defer chordNode.hashTable.RUnlock() //TODO can using array of pointers create a data race here?
+
+	fileRecords, doesExist := chordNode.hashTable.table[messageKeyPtr.Keyword]
+
+	if doesExist {
+		return &Val{FileRecords: fileRecords}, nil
+	} else {
+		return nil, errors.New(
+			fmt.Sprintf("such key does not exist at node IP %s ID %s", chordNode.node.Ip, chordNode.node.Ip))
```

Comment:
IP %s ID %s", chordNode.node.Ip, chordNode.node.**Ip** , should be 
 IP %s ID %s", chordNode.node.Ip, chordNode.node.**Id**